[1/n] Stabilize GCS/Autoscaler interface: Introduce monitor server

wuisawesome opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This is the first PR towards stabilizing the GCS autoscaler interface by introducing a new grpc service definition which we will provide backwards compatibility guarantees.

This PR mostly just introduces scaffolding and a trivial `GetRayVersion` endpoint. 

By the end of this series of PRs, `monitor.py` will only communicate with the rest of the ray cluster via this service definition. 

## Related issue number

Related to #31826

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

DmitriGekhtman commented on src/ray/protobuf/monitor.proto:
```diff
@@ -0,0 +1,30 @@
+  string version = 1;
+}
+
+// This service provides a stable interface for a monitor/autoscaler prococess to interact
```
```suggestion
// This service provides a stable interface for a monitor/autoscaler process to interact
```

gvspraveen commented on src/ray/gcs/gcs_server/gcs_monitor_server.h:
```diff
@@ -0,0 +1,21 @@
+namespace ray {
+namespace gcs {
+
+/// GcsNodeManager is responsible for managing and monitoring nodes as well as handing
```
replace `GcsNodeManager` with `GcsMonitorServer `

architkulkarni replied:
And the rest of the docstring too I suppose

gvspraveen reviewed (commented):
minor comment fix

rkooo567 commented:
Can you link me a design doc?

architkulkarni reviewed (approved):
Looks good, I'm assuming most of the gRPC code is boilerplate is copied from the existing GcsNodeManager/NodeInfoService, but let me know if there's something different here that warrants closer attention.

wuisawesome commented:
> Can you link me a design doc?

There have been no design changes since the previous design docs (you can refer to those design docs), I'm just refactoring everything to a cleaner defined protos. 

I've updated the list of things we need to handle in the github issue but there will be no design changes.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_server.cc:
```diff
@@ -598,6 +601,13 @@ void GcsServer::InitGcsTaskManager() {
   rpc_server_.RegisterService(*task_info_service_);
 }
 
+void GcsServer::InitMonitorServer() {
```
Any reason why we call it "MonitorServer" that's different from other modules' naming convention (*Manager)

wuisawesome replied:
Mostly because it's not actually "managing" anything. Most of the other managers actually manage some state.

wuisawesome commented:
Remaining test failures look like DinD issues and known flakey tests which are all unrelated. Merging.
